diff --git a/configs/common/FSConfig.py b/configs/common/FSConfig.py
index aa00efe..5c3bf06 100644
--- a/configs/common/FSConfig.py
+++ b/configs/common/FSConfig.py
@@ -81,7 +81,7 @@ def makeLinuxAlphaSystem(mem_mode, mdesc=None, ruby=False, cmdline=None):
 
     class BaseTsunami(Tsunami):
         ethernet = NSGigE(pci_bus=0, pci_dev=1, pci_func=0)
-        ide = IdeController(disks=[Parent.disk0, Parent.disk2],
+        ide = IdeController(disks=[Parent.disk0],
                             pci_func=0, pci_dev=0, pci_bus=0)
 
     self = LinuxAlphaSystem()
@@ -122,9 +122,7 @@ def makeLinuxAlphaSystem(mem_mode, mdesc=None, ruby=False, cmdline=None):
 
     self.mem_ranges = [AddrRange(mdesc.mem())]
     self.disk0 = CowIdeDisk(driveID='master')
-    self.disk2 = CowIdeDisk(driveID='master')
     self.disk0.childImage(mdesc.disk())
-    self.disk2.childImage(disk('linux-bigswap2.img'))
     self.simple_disk = SimpleDisk(disk=RawDiskImage(image_file = mdesc.disk(),
                                                read_only = True))
     self.intrctrl = IntrControl()
@@ -421,7 +419,7 @@ def makeArmSystem(mem_mode, machine_type, num_cpus=1, mdesc=None,
 def makeLinuxMipsSystem(mem_mode, mdesc=None, cmdline=None):
     class BaseMalta(Malta):
         ethernet = NSGigE(pci_bus=0, pci_dev=1, pci_func=0)
-        ide = IdeController(disks=[Parent.disk0, Parent.disk2],
+        ide = IdeController(disks=[Parent.disk0],
                             pci_func=0, pci_dev=0, pci_bus=0)
 
     self = LinuxMipsSystem()
@@ -436,9 +434,7 @@ def makeLinuxMipsSystem(mem_mode, mdesc=None, cmdline=None):
     self.bridge.master = self.iobus.slave
     self.bridge.slave = self.membus.master
     self.disk0 = CowIdeDisk(driveID='master')
-    self.disk2 = CowIdeDisk(driveID='master')
     self.disk0.childImage(mdesc.disk())
-    self.disk2.childImage(disk('linux-bigswap2.img'))
     self.malta = BaseMalta()
     self.malta.attachIO(self.iobus)
     self.malta.ide.pio = self.iobus.master
@@ -558,10 +554,8 @@ def makeX86System(mem_mode, numCPUs=1, mdesc=None, self=None, Ruby=False):
 
     # Disks
     disk0 = CowIdeDisk(driveID='master')
-    disk2 = CowIdeDisk(driveID='master')
     disk0.childImage(mdesc.disk())
-    disk2.childImage(disk('linux-bigswap2.img'))
-    self.pc.south_bridge.ide.disks = [disk0, disk2]
+    self.pc.south_bridge.ide.disks = [disk0]
 
     # Add in a Bios information structure.
     structures = [X86SMBiosBiosInformation()]
